### PR TITLE
[CMSIS-NN][Perf] Converted Relay Conv2D into CMSIS-NN Depthwise

### PR DESCRIPTION
**Pull request by ashutosh-arm**
This commit converts Relay Conv2D into depthwise
convolution. This is done only when total number
of output channels can be derived using the depth
multiplier. The tests include special case of output
channels=1.

cc @areusch

### CODE REVIEW

**Comment by ashutosh-arm** [comment]
cc @manupa-arm @Mousius @grant-arm for code review.

File: src/relay/backend/contrib/cmsisnn/relay_to_tir.cc [hunk 1]
```diff
@@ -31,6 +30,7 @@
 #include "../../../transforms/pattern_utils.h"
 #include "buffer_size.h"
 #include "compiler_attrs.h"
+#include "utils.h"
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply src/relay/backend/contrib/cmsisnn/relay_to_tir.cc]: Makes sense to me.

File: tests/python/contrib/test_cmsisnn/test_conv2d.py [hunk 2]
```diff
@@ -540,6 +542,111 @@ def test_depthwise_int8(
...
+        out_channels,
+        weight_format,
+        enable_bias,
+        relu_type,
+    )
+    orig_mod = make_module(model)
+    cmsisnn_mod = cmsisnn.partition_for_cmsisnn(orig_mod, params)
+
+    # validate pattern matching
+    assert_partitioned_function(orig_mod, cmsisnn_mod)
```
[the remark this answers is not in the document]
Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_conv2d.py:635]: I have added this check in the same test.

File: tests/python/contrib/test_cmsisnn/test_conv2d.py [hunk 3]
```diff
@@ -540,6 +547,135 @@ def test_depthwise_int8(
...
+    for target, mod in compiled_models[0].executor_factory.lowered_ir_mods.items():
+        if target.kind.name == "cmsis-nn":
+            cmsisnn_tir_mod = mod
+
+    cmsisnn_func = cmsisnn_tir_mod["tvmgen_default_cmsis_nn_main_0"]
+    call_extern = None
+    if isinstance(cmsisnn_func.body, tvm.tir.stmt.Evaluate):
+        call_extern = cmsisnn_func.body.value
+    else:
+        call_extern = cmsisnn_func.body.body.value
```
Comment (Mousius) [inline tests/python/contrib/test_cmsisnn/test_conv2d.py:666]: Isn't this stable between test runs?

Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_conv2d.py:666]: Its stable across runs. However, different parameters generate different TIR Primfunc.

Comment (ashutosh-arm) [reply tests/python/contrib/test_cmsisnn/test_conv2d.py:666]: To be specific, when the depth multiplier of Depthwise Convolution is 1, it does not allocate context buffer. This causes the Primfunc body to change.

File: tests/python/contrib/test_cmsisnn/test_conv2d.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mousius) [reply tests/python/contrib/test_cmsisnn/test_conv2d.py:635]: Ahh gotcha :+1:

File: tests/python/contrib/test_cmsisnn/test_conv2d.py (hunk 3 above)
Comment (Mousius) [reply tests/python/contrib/test_cmsisnn/test_conv2d.py:666]: Ahhh gotcha :+1:

**Review by Mousius — APPROVED** [review]
This is cool @ashutosh-arm 😸

**Pull request merged**